### PR TITLE
WIP: kepler exports CO2 intensity directly.

### PR DESCRIPTION
**Pull request by jbu**
This is a very rough sketch, but I thought it worth getting it out for comment.

The current method for viewing co2 metrics (https://github.com/sustainable-computing-io/kepler/discussions/600) is to combine Kepler-provided power metrics with public CO2 intensity APIs within prometheus.

It strikes me that 
* A large data centre might not be using the bog-standard power grid. It may have energy storage, or on-site generation, in which case it has a different carbon intensity.
* (slightly off topic for Kepler, but important down the track) It would be good to have a way for a hyperscaler to give co2-intensity of *services*. For instance, what is the co2 cost of a *message* in a particular region, or a GB of storage, 1s/vCPU of a lambda execution?

In which case the cluster itself is better placed to know and return it's carbon intensity.

So here's a simple metric that gets co2 intensity from the UK grid API and returns it as a stat. It doesn't try to then return different stats like co2/sec/cpu for dynamic, static, GPU, etc, etc. I'm not sure that would be fruitful.

Is this a useful thought? Comments?

### CODE REVIEW

File: pkg/co2/carbon_intensity.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+func Shutdown() bool {
+	if co2Impl != nil && config.EnabledCO2 {
+		return co2Impl.Shutdown()
+	}
+	return true
+}
+
+func GetCarbonIntensity(url string) (int64, error) {
+	if co2Impl != nil && config.EnabledCO2 {
+		return co2Impl.GetCarbonIntensity("https://api.carbonintensity.org.uk/intensity")
```
Comment (SamYuan1990) [inline pkg/co2/carbon_intensity.go:57]: May I have a brief for this API?

File: pkg/co2/source/ukgrid.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+func (d *UKGrid) Init() error {
+	d.collectionSupported = false
+	return nil
+}
+
+func (d *UKGrid) Shutdown() bool {
+	return true
+}
+
+func (d *UKGrid) GetCarbonIntensity(url string) (int64, error) {
```
Comment (rootfs) [inline pkg/co2/source/ukgrid.go:53]: @jbu can you also have a result cache here so the HTTP request happens less frequently?

**Comment by rootfs** [comment]
> Is this a useful thought? Comments?

yes it is :D

I think we can make CO2 metrics in two ways, inside (as this PR does) and outside (a separate service) of kepler. I am good both ways as long as the overhead is manageable. Here are my 2 cents around the overhead
### running inside Kepler
- Is the CO2 source query overhead manageable? If the service endpoint is reliable, this is not an issue. We have to have some ideas about what services to support.
- Is CO2 query scalable? Querying the source inside each Kepler would introduce many queries as the size of the cluster grows. However, if the query is cached, then this cost is still managable. 
- 
### running as a separate service
Kepler runs as it is, a separate deployment queries the CO2 source and emits yet another CO2 metrics to Prometheus. The end users combine both Kepler and CO2 metrics. This method shifts the query overhead to Prometheus, depending on the  runtime Prometheus configuration.

**Comment by marceloamaral** [comment]
Kepler runs as a Deamonset, so it creates a pod on all nodes. 
So, does it make sense make all Kepler instances report the same metrics?

Alternatively, could we use a separate exporter solely to export CO2 coefficients to Prometheus?

**Comment by marceloamaral** [comment]
Kepler runs as a Deamonset, so it creates a pod on all nodes. 
So, does it make sense make all Kepler instances report the same metrics?

Alternatively, could we use a separate exporter solely to export CO2 coefficients to Prometheus?

**Comment by williamcaban** [comment]
Questions:
- How does this approach account for the shift in power source types during the day which have different CO2e intensity? How will the cluster be aware of its CO2e intensity at any given time?
- A Tier3 or Tier4 datacenter has at least two independent power. How does this approach account for the situation when a datacenter uses two types of power sources with distinct CO2e intensity? How will the cluster calculate and determine the combination of these at any given time?
- (a permutation of the previous challenge) A large K8s cluster can span across multiple racks and these can have different power sources. How can the cluster be aware of this?

After thinking on these questions, could it be that we need an "on premise/local" API spec which can be used for more accurately understanding the CO2e intensity of the clusters and the nodes of the clusters in order to combine them with cluster & datacenter aware CO2e metrics?

File: pkg/co2/source/ukgrid.go (hunk 2 above)
Comment (jbu) [reply pkg/co2/source/ukgrid.go:53]: Definitely. As I wanted to spark discussion first I haven't done it. I'll add it when I get a chance.

File: pkg/co2/carbon_intensity.go (hunk 1 above)
Comment (jbu) [reply pkg/co2/carbon_intensity.go:57]: What sort of format are you looking for? I think this PR is more a discussion to design a brief.

**Comment by jbu** [comment]
To try and respond to all the comments: 
* It could be in Kepler or a separate source, yes.
* This API is agnostic regards how the co2 intensity is known or calculated. In this sketch I've used a public UK grid api. there are others. Which to use would depend on where the DC is among other things.
* Good question about what are we measuring? My initial thought is co2 intensity per service per power zone (?). If you have a cluster that spans power connections, what is our point of abstraction? It could be that internally k8s uses co2 aware scheduling, but since it's now making that internally it would publish some sort of average value for a vcpu-second?

File: pkg/co2/carbon_intensity.go (hunk 1 above)
Comment (SamYuan1990) [reply pkg/co2/carbon_intensity.go:57]: @jbu , when I access this API, it responses as below:
```
{ 
  "data":[{ 
    "from": "2023-05-14T11:00Z",
    "to": "2023-05-14T11:30Z",
    "intensity": {
      "forecast": 121,
      "actual": 112,
      "index": "low"
    }
  }]
}
```
Hence serval questions:
1. what does this API mean?
2. what does this json format means?
3. according to `org.uk` ? so it's some data from which org? and considering other regions of world, any other endpoints?
etc....
you know, for any one without any background knowledge about this API to review this part of code, there are lots of question marks in their mind. cc: @rootfs

File: pkg/co2/source/ukgrid.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	}
+
+	var intensity Intensity
+	err = json.Unmarshal(body, &intensity)
+	if err != nil {
+		fmt.Println(err)
+		return 0, err
+	}
+
+	fmt.Printf("%+v\n", intensity)
```
Comment (SamYuan1990) [inline pkg/co2/source/ukgrid.go:80]: avoid fmt.Printf as it seems escapes to heap and if it's just log, please use log.

**Comment by SamYuan1990** [comment]
> To try and respond to all the comments:
> 
> * It could be in Kepler or a separate source, yes.
> * This API is agnostic regards how the co2 intensity is known or calculated. In this sketch I've used a public UK grid api. there are others. Which to use would depend on where the DC is among other things.
> * Good question about what are we measuring? My initial thought is co2 intensity per service per power zone (?). If you have a cluster that spans power connections, what is our point of abstraction? It could be that internally k8s uses co2 aware scheduling, but since it's now making that internally it would publish some sort of average value for a vcpu-second?

here is my personal point of view. 
I suppose this PR is try to provide an approach as a rate for carbon and use this rate multiplication power consumption to get the specific carbon consumption pre each pod? 
It's interesting that if kepler can inject with external carbon data. 
if I am right, which means kepler can report carbon consumption directly.

At carbon/business side
In real world, there serval ways to convert other power to electrical. Some product carbon, some are not（Solar power？）.
Hence 
1. for a specific data center, what's the meaning to use an avg data among region as standard or not?
2. for data center or edge node, we can't use data for UK to cover all the case, right?
... lots of cases and questions out of my knowledge, hence I hope some reference can be listed here in this PR as background supporting.

At architecture side
when we use kepler with edge device, maybe we don't need this feature.
basing on my knowledge at today, maybe it's better to be an external service, or I would like put this kind of integration works at grafana?
I am not sure if kepler can support a post way to receive those data instead of inside implementation to get the data?
as the current endpoints in this PR is not following any standard(please correct me for any mistake)
considering extend kepler to support any other API endpoints.

**Comment by rootfs** [comment]
@jbu we'll have separate exporters for external power sources. CO2 metrics is a good fit, would you like to move to its own repo under `sustainable-computing-io`? Additionally, have you checked [Carbon aware SDK](https://github.com/Green-Software-Foundation/carbon-aware-sdk)? It'll be great if the SDK can be reused to some extents.

**Comment by ajgillette** [comment]
I'm trying to install kepler on a linux KinD cluster and the kepler-exporter is not connecting to the ebpf resource. The error:

modprobe: FATAL: Module kheaders not found in directory /lib/modules/5.15.49-linuxkit-pr
chdir(/lib/modules/5.15.49-linuxkit-pr/build): No such file or directory
I0605 17:34:25.814687       1 bcc_attacher.go:73] failed to attach the bpf program: <nil>

**Comment by ajgillette** [comment]
@rootfs I'm running on native Linux (Ubuntu 22.04) I tried both helm install and building from source. Same error. I did an exec to the running pod and was able to verify that /lib/modules/5.15.49-linuxkit-pr/build does exist inside the pod.

**Comment by ajgillette** [comment]
@rootfs: The recipe to build the source is:

git clone https://github.com/sustainable-computing-io/kepler.git
make build-manifest OPTS=PROMETHEUS_DEPLOY
kubectl apply -f _output/generated-manifest/deployment.yaml

**Comment by ajgillette** [comment]
@rootfs: I am running kubeshark on this KinD cluster so I know that it is possible to connect to bpf modules.

**Comment by rootfs** [comment]
@ajgillette do you have this `/usr/src` mount config to bind mount `/usr/src` to your kind cluster and then pass to kepler? https://github.com/sustainable-computing-io/local-dev-cluster/blob/main/kind/manifests/kind.yml

**Comment by ajgillette** [comment]
When I exec to the kepler-exporter these are the only kepler metrics exposed:

curl localhost:9102/metrics | grep kepler
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  7738    0  7738    0     0  1259k      0 --:--:-- --:--:-- --:--:-- 1259k
# HELP kepler_exporter_build_info A metric with a constant '1' value labeled by version, revision, branch, and goversion from which kepler_exporter was built.
# TYPE kepler_exporter_build_info gauge
kepler_exporter_build_info{branch="",goversion="go1.18.1",revision="",version=""} 1
# HELP kepler_node_energy_stat Several labeled node metrics
# TYPE kepler_node_energy_stat counter
kepler_node_energy_stat{cpu_architecture="unknown",node_block_devices_used="0",node_curr_bytes_read="0",node_curr_bytes_writes="0",node_curr_cache_miss="0",node_curr_container_cpu_usage_seconds_total="0",node_curr_container_memory_working_set_bytes="0",node_curr_cpu_cycles="0",node_curr_cpu_instr="0",node_curr_cpu_time="0",node_curr_energy_in_core_joule="0",node_curr_energy_in_dram_joule="0",node_curr_energy_in_gpu_joule="0",node_curr_energy_in_other_joule="0",node_curr_energy_in_pkg_joule="0",node_curr_energy_in_uncore_joule="0",node_name="kepler-exporter-m28fx"} 0
# HELP kepler_node_nodeInfo Labeled node information
# TYPE kepler_node_nodeInfo counter
kepler_node_nodeInfo{cpu_architecture="unknown"} 1
# HELP kepler_node_other_host_components_joules_total Aggregated RAPL value in other components (platform - package - dram) in joules
# TYPE kepler_node_other_host_components_joules_total counter
kepler_node_other_host_components_joules_total{instance="kepler-exporter-m28fx",mode="dynamic"} 0
kepler_node_other_host_components_joules_total{instance="kepler-exporter-m28fx",mode="idle"} 0
# HELP kepler_node_platform_joules_total Aggregated RAPL value in platform (entire node) in joules
# TYPE kepler_node_platform_joules_total counter
kepler_node_platform_joules_total{instance="kepler-exporter-m28fx",mode="dynamic",source="acpi"} 0
kepler_node_platform_joules_total{instance="kepler-exporter-m28fx",mode="idle",source="acpi"} 0

**Comment by ajgillette** [comment]
my KinD cluster is running on a linux server:

ajgillette@FredsDesktop:/etc$ uname -r
5.15.0-72-generic
ajgillette@FredsDesktop:/etc$ cat /proc/version
Linux version 5.15.0-72-generic (buildd@lcy02-amd64-035) (gcc (Ubuntu 11.3.0-1ubuntu1~22.04.1) 11.3.0, GNU ld (GNU Binutils for Ubuntu) 2.38) #79-Ubuntu SMP Wed Apr 19 08:22:18 UTC 2023

**Comment by rootfs** [comment]
@ajgillette can you post the kernel source directory on the host?
```console
ls -l /lib/modules/5.15.49-linuxkit-pr/build
```

**Comment by ajgillette** [comment]
So the kepler-exporter has a path/link -> "/lib/modules/5.15.49-linuxkit-pr/build -> /usr/src/linux-headers-5.15.49-linuxkit-pr" but that directory doesn't actually exist in the container. If I cd to "/usr/src" there are two directories, one called "debug" and one called "kernel". There is no directory "linux-headers-5.15.49-linuxkit-pr" . If I cd into /usr/src/kernel, there is a directory called "4.18.0-486.el8.x86_64" and if I grep -R the kernel directory there is not linuxkit. Since this is inside the kepler-exporter pod, I'm assuming that the pod was not built with the correct paths? 

Is there some additional config that I have to do to build the pod properly?

**Comment by ajgillette** [comment]
My current pod as viewed from Kubeshark:
ks

**Comment by ajgillette** [comment]
Sorry that should have said cluster no pod

**Comment by rootfs** [comment]
@ajgillette can you stand up kind through [this](https://github.com/sustainable-computing-io/local-dev-cluster) instruction?

**Comment by ajgillette** [comment]
I checked out the files and created a cluster using the instruction you suggested but it's still doing the same thing.  What do you mean when you say "Locate your BCC lib and linux header " and "modify kind config to make sure extraMounts: cover linux header and BCC."

**Comment by rootfs** [comment]
here is the workflow of deploying bcc and kind. https://github.com/sustainable-computing-io/kepler-action/blob/main/index.js

**Comment by ajgillette** [comment]
@rootfs Thanks for all your help so far, but I still don't have Kepler working. The docs say to make sure that mounts are added for BCC and Linux headers but they don't say how to do that. I have figured out that the missing headers are from a linuxkit kernel and I found the kernel in question but I haven't figured out how to install them. I can get them in a container but that's not very helpful. I checked out the project with the workflow that you suggested but I'm not sure how to use it. I've watched all the posted videos and I'm going to try to do a talk tomorrow morning at 8:30 EDST but a demo is not looking good. Any last minute suggestions?

**Comment by rootfs** [comment]
@ajgillette have you tried minikube? minikube has all the ebpf support. You can follow the process of installing kepler. I just did it on my RHEL 8 and it works

https://gist.github.com/rootfs/f6338b29a10808064caeeecc4ff6121a

**Comment by ajgillette** [comment]
@rootfs I got it all working last night, but it was a non-trivial. Is there a way that we could connect in real time. I work for Nokia and we are looking at a number of different ESG related projects. I was asked to evaluate Kepler and I have submitted my work on that. I did a presentation this morning that went over well. Now I am being asked to look at PEAKS and I would like to have your thoughts on the best was to show it in action. My work email is fred.gillette@nokia.com. Maybe connect on LinkedIn?

**Comment by rootfs** [comment]
@ajgillette added you to slack

**Comment by ajgillette** [comment]
ok I'm relatively new to slack, let me see if I can find it. My linkedIn profile is https://www.linkedin.com/in/fredgillette/

**Pull request closed without merging**